### PR TITLE
dnsdist: Implement a 'lazy' health-checking mode

### PR DESCRIPTION
**Pull request by rgacogne**
The general idea is that, on some low-end devices, we want to avoid sending health-check queries to the backend at a regular interval, as this might represent an order of magnitude more queries than regular traffic.
To do that we instead look at the amount of timeouts and Server Failure errors occurring with regular queries, and only start doing active health-checking if these errors reach a certain threshold, temporarily placing the backend in a 'potential failure' state. If the health-check queries succeed, we quickly go back to the previous 'Healthy' state and stop sending health-check queries. Otherwise we place the backend in 'Failed' state and keep sending health-check queries but a reduced rate, as long as the backend is failing.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

File: pdns/dnsdistdist/dnsdist-backend.cc [hunk 1]
```diff
@@ -475,6 +497,128 @@ IDState* DownstreamState::getIDState(unsigned int& selectedID, int64_t& generati
...
+    return;
+  }
+
+  if (newState) {
+    /* check succeeded */
+    currentCheckFailures = 0;
+
+    if (!upStatus) {
+      /* we were marked as down */
+      consecutiveSuccessfulChecks++;
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdistdist/dnsdist-backend.cc:583]: We are still in the 'down' state but just had a successful health-check result.

File: pdns/dnsdistdist/test-dnsdistbackend_cc.cc [hunk 2]
```diff
@@ -179,7 +179,7 @@ BOOST_AUTO_TEST_CASE(test_Lazy)
   BOOST_CHECK_EQUAL(ds.healthCheckRequired(), true);
 
   /* we need maxCheckFailures failed health-checks to go down */
-  for (size_t idx = 0; idx < config.maxCheckFailures - 1; idx++) {
+  for (size_t idx = 0; idx < static_cast<size_t>(config.maxCheckFailures - 1); idx++) {
```
Comment (cmouse) [inline pdns/dnsdistdist/test-dnsdistbackend_cc.cc:182]: this will roll for a while if maxCheckFailures == 0

Comment (rgacogne) [reply pdns/dnsdistdist/test-dnsdistbackend_cc.cc:182]: It's good that we explicit set `config.maxCheckFailures = 3;` above, then :)

File: pdns/dnsdistdist/dnsdist-backend.cc [hunk 3]
```diff
@@ -475,6 +499,200 @@ IDState* DownstreamState::getIDState(unsigned int& selectedID, int64_t& generati
...
+        if (result) {
+          ++failures;
+        }
+      }
+
+      const auto maxFailureRate = static_cast<float>(d_config.d_lazyHealthChecksThreshold);
+      auto current = (100.0 * failures) / totalCount;
+      if (current >= maxFailureRate) {
+        lastResults.clear();
+        vinfolog("Backend %s reached the lazy health-check threshold (%f out of %f, looking at sample of %d items with %d failures), moving to Potential Failure state", getNameWithAddr(), current, maxFailureRate, totalCount, failures);
```
Comment (Habbie) [inline pdns/dnsdistdist/dnsdist-backend.cc:540]: perhaps append `%` to the `current` and `maxFailureRate` values

File: pdns/dnsdist-lua.cc [hunk 4]
```diff
@@ -507,6 +526,57 @@ static void setupLuaConfig(LuaContext& luaCtx, bool client, bool configCheck)
...
+                         if (vars.count("lazyHealthCheckMinSampleCount")) {
+                           auto value = std::stoi(boost::get<string>(vars.at("lazyHealthCheckMinSampleCount")));
+                           checkParameterBound("lazyHealthCheckMinSampleCount", value);
+                           config.d_lazyHealthChecksMinSampleCount = value;
+                         }
+
+                         if (vars.count("lazyHealthCheckThreshold")) {
+                           auto value = std::stoi(boost::get<string>(vars.at("lazyHealthCheckThreshold")));
+                           checkParameterBound("lazyHealthCheckThreshold", value, std::numeric_limits<uint8_t>::max());
+                           config.d_lazyHealthChecksThreshold = value;
```
Comment (Habbie) [inline pdns/dnsdist-lua.cc:544]: `lazyHealthCheckThreshold` being stored in `d_lazyHealthChecksThreshold` looks a bit weird (the extra s)

**Comment by rgacogne** [comment]
Fixed the issues from Peter's review, and rebased to fix a conflict.

**Review by cmouse — APPROVED** [review]

**Pull request merged**